### PR TITLE
Changes for MonetDBJavaLite compilation

### PR DESCRIPTION
**Pull request by PedroTadim**
Here are all my changes for the native code compilation for MonetDBJavaLite. Any question please ask. Thanks!

### CODE REVIEW

File: src/common/mutils.h [hunk 1]
```diff
@@ -40,9 +40,9 @@
 
 mutils_export int MT_lockf(char *filename, int mode, off_t off, off_t len);
 
-#if (defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__
-extern int fsync (int fd);
-extern int winerror(int e);
+#if ((defined _WIN32 || defined __WIN32__) && ! defined __CYGWIN__) || (defined(HAVE_EMBEDDED_JAVA) && defined(NATIVE_WIN32))
+mutils_export int fsync (int fd);
```
Comment (hannes) [inline src/common/mutils.h]: Do you depend on mutils_export?

File: src/common/stream.c [hunk 2]
```diff
@@ -168,7 +145,9 @@ struct stream {
 	union {
 		void *p;
 		int i;
+#if !defined(HAVE_EMBEDDED_JAVA) || !defined(NATIVE_WIN32)
 		SOCKET s;
```
Comment (hannes) [inline src/common/stream.c]: WHat do you need the SOCKET for? We can remove it...

File: src/common/stream.h [hunk 3]
```diff
@@ -93,6 +93,28 @@ typedef __int128_t hge;
 #define STREAM_FWF_RECORD_SEP '\n'
 #define STREAM_FWF_FILLER ' '
 
+#ifdef _MSC_VER
```
Comment (hannes) [inline src/common/stream.h:95]: What's the point of moving this to the header?

File: src/embedded/embedded.c [hunk 4]
```diff
@@ -270,17 +271,34 @@ char* monetdb_query(monetdb_connection conn, char* query, char execute, monetdb_
 		goto cleanup;
 	}
 
-	if (!m->results && m->rowcnt >= 0 && affected_rows) {
+	if (affected_rows) {
 		*affected_rows = m->rowcnt;
 	}
 
+#ifdef HAVE_EMBEDDED_JAVA
```
Comment (hannes) [inline src/embedded/embedded.c]: Is this really specific to embedded Java? Can't this be done in a generic way?

File: src/embedded/embedded.h [hunk 5]
```diff
@@ -13,7 +13,7 @@
 #ifndef _EMBEDDED_LIB_
 #define _EMBEDDED_LIB_
 
-#include <stddef.h> /* only for size_t */
+#include "monetdb_config.h"
```
Comment (hannes) [inline src/embedded/embedded.h]: Nope. I don't want any MonetDB headers in embedded.h.

File: src/gdk/gdk_bat.c [hunk 6]
```diff
@@ -374,6 +383,43 @@ BATattach(int tt, const char *heapfile, int role)
 		p = Tloc(bn, 0);
 		n = (lng) st.st_size;
 		while (n > 0 && (m = fread(p, 1, (size_t) MIN(1024*1024, n), f)) > 0) {
+#if defined(HAVE_EMBEDDED) && defined(HAVE_EMBEDDED_JAVA)
```
Comment (hannes) [inline src/gdk/gdk_bat.c:378]: Why swap? Is the endianness in the JVM different?

File: src/gdk/gdk_bbp.c [hunk 7]
```diff
@@ -1518,7 +1517,7 @@ BBPexit(void)
 	/* free all memory (just for leak-checking in Purify) */
 	do {
 		skipped = 0;
-		for (i = 0; i < (bat) ATOMIC_GET(BBPsize, BBPsizeLock); i++) {
+		for (i = 0; i < (bat) BBPsize; i++) {
```
Comment (hannes) [inline src/gdk/gdk_bbp.c]: I suppose there is a reason for this lock. Why remove?

File: src/gdk/gdk_posix.c [hunk 8]
```diff
@@ -1041,6 +1045,44 @@ MT_sleep_ms(unsigned int ms)
...
+		freeMonetDB5LibraryPath();
+	}
+	monetdb_lib_path = GDKstrdup(path);
+	return monetdb_lib_path != NULL ? 1 : 0;
+}
+
+#ifdef NATIVE_WIN32
+/* on Windows compilation of MonetDBJavaLite, this function was still missing */
+/* adapted from https://llvm.org/svn/llvm-project/compiler-rt/branches/release_30/lib/popcountdi2.c */
+int __popcountdi2(unsigned long long a) {
```
Comment (hannes) [inline src/gdk/gdk_posix.c]: what is it used for?

File: src/mal/mal/mal_module.c [hunk 9]
```diff
@@ -28,6 +28,7 @@
  */
```
Comment (hannes) [inline src/mal/mal/mal_module.c:31]: where do these changes come from? default?

File: src/mal/sqlbackend/sql.c [hunk 10]
```diff
@@ -1943,19 +1946,56 @@ mvc_export_table_wrap( Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci)
...
-				      filename?filename:"stdout", strerror(errnr));
-		goto wrapup_result_set1;
+	else {
+		flen = strlen(filename);
+		if ((fn = GDKmalloc(flen + 1)) == NULL) {
+			msg = createException(SQL, "sql.resultSet", MAL_MALLOC_FAIL);
+			goto wrapup_result_set1;
+		}
+#if defined(HAVE_EMBEDDED) && defined(WIN32)
+		// fix single backslash file separator on windows
```
Comment (hannes) [inline src/mal/sqlbackend/sql.c:2098]: thats the code to make copy into work?

File: src/mal/sqlbackend/sql_statement.c [hunk 11]
```diff
@@ -124,15 +124,32 @@ stmt_key(stmt *s)
 }
 
 /* #TODO make proper traversal operations */
+#ifdef HAVE_EMBEDDED
 stmt *
-stmt_atom_string(backend *be, const char *S)
+stmt_atom_string(backend *be, const char *S, int escape)
+{
+	const char *s = NULL;
```
Comment (hannes) [inline src/mal/sqlbackend/sql_statement.c:131]: again, changes from default??

**Review by hannes — CHANGES_REQUESTED** [review]
here are some detailed comments. Is this really a minimal set of changes?

File: src/common/mutils.h (hunk 1 above)
Comment (PedroTadim) [reply src/common/mutils.h]: I was using it to keep the extern keyword like the other declarations in that header file, but you can replace it the extern word instead.

File: src/common/stream.c (hunk 2 above)
Comment (PedroTadim) [reply src/common/stream.c]: Nothing! I was getting a compilation error on Windows. We can try to remove most of the content of the stream.c module.

File: src/common/stream.h (hunk 3 above)
Comment (PedroTadim) [reply src/common/stream.h:95]: Maybe I can move it to the implementation file. Yes!

File: src/embedded/embedded.h (hunk 5 above)
Comment (PedroTadim) [reply src/embedded/embedded.h]: It was for the Windows compilation. I will try to get rid of it.

File: src/embedded/embedded.c (hunk 4 above)
Comment (PedroTadim) [reply src/embedded/embedded.c]: For the JDBC connection is tricky, but I can look further. Changing the monetdb_query function signature is probably a bad idea due to the large number of parameters.

File: src/gdk/gdk_bat.c (hunk 6 above)
Comment (PedroTadim) [reply src/gdk/gdk_bat.c:378]: The JVM is always big endian. My machine is little endian and I was getting problems with the binary import.

File: src/gdk/gdk_bbp.c (hunk 7 above)
Comment (PedroTadim) [reply src/gdk/gdk_bbp.c]: I am locking the whole function. I was getting a race condition while shutting down the embedded database in one of the stress tests, so I had to lock the whole of it.

File: src/gdk/gdk_posix.c (hunk 8 above)
Comment (PedroTadim) [reply src/gdk/gdk_posix.c]: That symbol was missing in the Windows DLL.

File: src/mal/mal/mal_module.c (hunk 9 above)
Comment (PedroTadim) [reply src/mal/mal/mal_module.c:31]: Mine in MonetDBJavaLite. I was having a race condition there while shutting down the database in one of the stress tests.

File: src/mal/sqlbackend/sql.c (hunk 10 above)
Comment (PedroTadim) [reply src/mal/sqlbackend/sql.c:2098]: Yes

File: src/mal/sqlbackend/sql_statement.c (hunk 11 above)
Comment (PedroTadim) [reply src/mal/sqlbackend/sql_statement.c:131]: Mine. The CSV export was still having problems on Windows. The file path was getting manipulated by the sql2str call, thus failing the export.

File: src/common/stream.h (hunk 3 above)
Comment (PedroTadim) [reply src/common/stream.h:95]: I guess from the API point of view it's more organized

File: src/gdk/gdk_posix.c (hunk 8 above)
Comment (PedroTadim) [reply src/gdk/gdk_posix.c]: Fixed it. Using a compiler flag.

File: src/common/stream.c (hunk 2 above)
Comment (PedroTadim) [reply src/common/stream.c]: Fixed (removed it).

**Comment by hannes** [comment]
Please make sure the travis build passes with the changes in this PR

**Comment by hannes** [comment]
Please merge again, also note the trick in the Makefile to avoid the pthreads dependency

**Comment by PedroTadim** [comment]
I have merged the changes, but I get a problem on linking with your flags. I am distributing the libwinpthread-1.dll, and I am having no problems. I have just fixed all the points.

**Comment by hannes** [comment]
Great that it passes now. I have one more request: Every change creates a liability for future merges, so could you check these changes are minimal?

**Comment by hannes** [comment]
Also, whats the problem when linking?!

**Comment by PedroTadim** [comment]
Actually I had the mingw64-winpthreads-static package missing :), so that's fixed. The only thing left to fix is the monetdb_query result. I simplified it, but I guess you want the query result type passed as a parameter.

**Pull request closed without merging**